numba/core/transforms.py

```diff
@@ -477,73 +476,130 @@ def _cfg_nodes_in_region(cfg, region_begin, region_end):
+            return isinstance(stmt, ir.Branch)
+
+
+        def is_terminator(stmt):
+            return isinstance(stmt, ir.Terminator)
+
+        def is_raise(stmt):
+            return isinstance(stmt, ir.Raise)
+
+        def is_return(stmt):
+            return isinstance(stmt, ir.Return)
+
+        def is_pop_block(stmt):
+            try:
+                if hasattr(stmt, "value"):
+                    return str(stmt.value).startswith("POP_BLOCK_INFO")
+            except KeyError:
+                return False
+
+        cfg = compute_cfg_from_blocks(blocks)
+        sus_setups, sus_pops = [], []
+        for label, block in blocks.items():
+            for stmt in block.body:
+                if is_setup_with(stmt):
+                    sus_setups.append(label)
+                if is_pop_block(stmt):
+                    sus_pops.append(label)
+
+        # now that we do have the statements, iterate through them in reverse
+        # topo order and from each start looking for pop_blocks
+        setup_with_to_pop_blocks_map = defaultdict(set)
+        for setup_block in cfg.topo_sort(sus_setups, reverse=True):
+            # begin pop_block, search
+            to_visit, seen = [], []
+            to_visit.append(setup_block)
+            while to_visit:
+                # get whatever is next and record that we have seen it
+                block = to_visit.pop()
+                seen.append(block)
+                # go through the body of the block, looking for statements
+                for stmt in blocks[block].body:
+                    # raise detected before pop_block
+                    if is_raise(stmt):
+                            raise errors.CompilerError(
+                                'unsupported controlflow due to raise '
+                                'statements inside with block'
+                                )
+                    # special case 3.7, return before POP_BLOCK
+                    if PYVERSION < (3, 8) and is_return(stmt):
+                            raise errors.CompilerError(
+                                'unsupported controlflow due to return '
+                                'statements inside with block'
+                                )
+                    # if a pop_block, process it
+                    if is_pop_block(stmt) and block in sus_pops:
+                        # record the jump target of this block belonging to this setup
+                        pop_block_targets = blocks[block].terminator.get_targets()
+                        if len(pop_block_targets) != 1:
+                            raise errors.CompilerError(
+                                "Does not support with-context that contain branches "
+                                "(i.e. break/return/raise) that can leave the with-context. "
+                            )
+                        target_block = blocks[pop_block_targets[0]]
+                        if not target_block.terminator.get_targets():
```

Comment:
Would it be clearer to check if the target block terminator is an instance of an `ir.Return`, i.e.
```python
if is_return(target_block.terminator):
    ...
```
?

Reply:
done in a1eb14a0f2f044623366036177af991cc9528a4d